plugins/modules/source_control/gitlab/gitlab_runner.py

```diff
@@ -343,6 +345,7 @@ def main():
             ['api_username', 'api_job_token'],
             ['api_token', 'api_oauth_token'],
             ['api_token', 'api_job_token'],
+            ['project', 'owned'],
```

Comment:
Do we know for sure that this always results in an API error with every version of GitLab and the Python GitLab module? (Or can we reasonably assume it always does?)

Reply:
When you set a value to `project`, and `owned=false` (the usecase we want to avoid here), the gitlab_runner module calls `self._runners_endpoint.all()`, which translates to `ProjectRunnerManager.all()` in python-gitlab library.

This call does not exists in recent versions of the library, and `/projects/:id/runners/all` (which is behind the `all()`) has never been documented in GitLab API resources documentation since 10.3 to current 14.8-pre.

Perhaps it existed one day, perhaps python-gitlab library implemented it one day, but I see no reason for it to exist.
@nejch may you enlighten us on this subject ?

Reply:
Thanks for reaching out @lgatellier! You're right that `ProjectRunnerManager.all()` does not exist and has no reason to. And except for a small misleading snippet in GitLab's API docs, `/projects/:id/runners/all` also does not exist (you'll get a 404) and never did, looking at the original MR in [GitLab 8.4](https://gitlab.com/gitlab-org/gitlab-foss/-/merge_requests/2640/diffs) when it was still API v3 even. (The misleading docs were actually introduced _today_ about 10 hours ago: https://gitlab.com/gitlab-org/gitlab/-/commit/d493fea4927329758424e662c30b5ef2de0d0642)

In general, assuming `gl = gitlab.Gitlab()` and `project = gl.projects.get(project_id)`:

- `gl.runners.list()` calls `/api/v4/runners`, returns **specific runners** available to the authenticated users(not sure if owned is the best term, but GitLab used it)
- `gl.runners.all()` calls `/api/v4/runners/all`, returns **shared and specific runners** on the instance (admin-only endpoint)
- `project.runners.list()` calls `/api/v4/projects/:id/runners`, returns **shared and specific runners** available to the project

So I also agree that `owned` makes no sense in the context of a project.

BTW: all of the remaining `all()` methods in python-gitlab are problematic and also why you have to have [this hack](https://github.com/ansible-collections/community.general/blob/b444dc81a127d96f9e1560b3932237ccbcaf1c66/plugins/modules/source_control/gitlab/gitlab_runner.py#L291-L298) in the module.

I plan to introduce a separate manager for `runners_all.list()` instead for the admin endpoint (as we already did for `members_all.list()`) and that should make things clearer. I could do this soon and make a new release, but I guess you'll want to keep compatibility with older `all()` versions.